yesod-core/Yesod/Core/Handler.hs

```diff
@@ -787,6 +789,30 @@ setHeader :: MonadHandler m => Text -> Text -> m ()
 setHeader = addHeader
 {-# DEPRECATED setHeader "Please use addHeader instead" #-}
 
+-- | Replace an existing header with a new value or add a new header
+-- if not present.
+--
+-- Note that, while the data type used here is 'Text', you must provide only
+-- ASCII value to be HTTP compliant.
+--
+-- @since 1.4.36
+replaceOrAddHeader :: MonadHandler m => Text -> Text -> m ()
+replaceOrAddHeader a b =
+  let header = Header (encodeUtf8 a) (encodeUtf8 b)
+  in modify $ \g -> g {ghsHeaders = replaceHeader header (ghsHeaders g)}
+  where
+    sameHeaderName :: Header -> Header -> Bool
+    sameHeaderName (Header n1 _) (Header n2 _) = n1 == n2
+    sameHeaderName _ _ = False
+
+    replaceHeader :: Header -> Endo [Header] -> Endo [Header]
+    replaceHeader header endo =
+      let allHeaders :: [Header] = appEndo endo []
+      in Endo
```

Comment:
Hmm.. It looks as if the ordering is still wrong in your latest repo.  I am trying this

```
addHeader "Foo" "Bar"
replaceOrAddHeader "Foo" "Baz"
addHeader "Foo" "Dup"
replaceOrAddHeader "FOO" "Final"
```

and I get

```
FOO: Final
Foo: Baz
Vary: Accept, Accept-Language
Foo: Dup
```

I did have a quick look at this earlier in the week, and I ended up with a modified version of `tell` which also takes a filtering function `alter` wrapped up in `Endo`:
```
alterAndTell alter hs = modify f
  where f g = let altered = Endo (appEndo (alter `mappend` ghsHeaders g) [] ++)
              in g { ghsHeaders = altered `mappend` hs }
```